src/pytorch_lightning/core/module.py

```diff
@@ -1429,7 +1466,8 @@ def toggle_optimizer(self, optimizer: Union[Optimizer, LightningOptimizer], opti
         # Iterate over all optimizer parameters to preserve their `requires_grad` information
         # in case these are pre-defined during `configure_optimizers`
         param_requires_grad_state = {}
-        for opt in self.trainer.optimizers:
+        optimizers = self._optimizers if self._fabric is not None else self.trainer.optimizers
+        for opt in optimizers:
```

Comment:
Couldn't we do this via the shim? Is `Fabric.optimizers` not useful?

Even if we don't need to hold an owning reference, it could be a weak reference.

Reply:
> Is Fabric.optimizers not useful?

I want to avoid keeping track of modules and optimizers inside Fabric. Fabric has a very "functional" approach and this makes it flexible, and at the same time easy to understand. Unless it can improve the UX, I vote for not doing that.

Reply:
I agree, let's not do that for now, we can always change our minds later but let's keep it leaner for now.